Add better hint in the doctor for Maven workspaces (#744)
scalameta/metals: docs/build-tools/maven.md

---
id: maven
title: Maven
---

> Maven is one of the most common build tools in the JVM ecosystem and it also allows for using scala through the [scala-maven-plugin](https://davidb.github.io/scala-maven-plugin/usage.html).

Reviewer: ```suggestion
Maven is one of the most common build tools in the JVM ecosystem and it also allows for using Scala through [scala-maven-plugin](https://davidb.github.io/scala-maven-plugin/usage.html) and [scalor-maven-plugin](https://github.com/random-maven/scalor-maven-plugin).
```

Author: I don't think bloop plugin works with scalor, so maybe we shouldn't mention it?

Reviewer: I didn't know. Is there a something we can link to for scalor-maven-plugin users? It seems like an actively maintained project with fairly rich features. If so, we can add a new sentence
```
also allows for using scala through the [scala-maven-plugin](https://davidb.github.io/scala-maven-plugin/usage.html).
The [scalor-maven-plugin](https://github.com/random-maven/scalor-maven-plugin) is not supported, see $LINK for more information.
```

Author: > I didn't know. Is there a something we can link to for scalor-maven-plugin users?

I don't think we have anything to point to right now - maybe add a comment that we would appreciate a contribution to bloop?

Reviewer: Fair enough, we can skip mentioning scalor-maven-plugin in that case and if somebody asks we'll point to Bloop.

## Automatic installation

The first time you open Metals in a new workspace it prompts you to import the
build. Select "Import build" to start automatic installation.

> This will create all the needed bloop files, however there will be a warning, since the semanticDB plugin was not added yet in the automatic import. This will be added later. Most features should work without it, however some require semanticDB files to be provided alongside compiled data. To do that we need a couple of steps that are explained in the manual instalation section.

Reviewer: I recommend formatting markdown files with prettier even if we don't enforce this in CI. Prettier provides automatic line-wrapping which makes reviewing a bit easier.

Reviewer: ```suggestion
This will create all the needed Bloop files, however there will be a warning, since the SemanticDB plugin was not added yet in the automatic import. This will be added later. Most features should work without it, however some require SemanticDB files to be provided alongside compiled data. To do that we need a couple of steps that are explained in the manual installation section.
```

## Manual instalation

> First, we need to add a couple of options to the scala compiler in the configuration section:

Reviewer: ```suggestion
First, we need to add a couple of options to the Scala compiler in the configuration section:
```

```xml
<plugin>
<groupId>net.alchim31.maven</groupId>
<artifactId>scala-maven-plugin</artifactId>
<version>4.0.2</version>
...
<configuration>
    <compilerPlugins>
        <compilerPlugin>
            <groupId>org.scalameta</groupId>
            <artifactId>semanticdb-scalac_${scala.version}</artifactId>
            <version>@@SCALAMETA_VERSION@@</version>
        </compilerPlugin>
    </compilerPlugins>
    <args>
        <arg>-P:semanticdb:synthetics:on</arg>
        <arg>-P:semanticdb:failures:warning</arg>
        <arg>-P:semanticdb:sourceroot:${project.basedir}</arg>
        <arg>-Yrangepos</arg>
        <arg>-Xplugin-require:semanticdb</arg>
    </args>
</configuration>
...
</plugin>
```

> `@@SCALAMETA_VERSION@@` version can be replaced with whatever is the newest semanticDB version.

Reviewer: ```suggestion

```

Not needed, `@@SCALAMETA_VERSION@@` is automatically replaced by mdoc.

Next, we need to do run bloopInstall via maven, which can be done easily through:

> `mvn ch.epfl.scala:maven-bloop_2.10:1.3.0:bloopInstall -DdownloadSources=true`

Reviewer: ```suggestion
`mvn ch.epfl.scala:maven-bloop_2.10:@@BLOOP_VERSION@@:bloopInstall -DdownloadSources=true`
```

> where `1.3.0` can be replaced with newer bloop version.

Reviewer: ```suggestion
```

Everything should now be correctly configured and work even when reimporting the project.

If you don't want to modify the `pom.xml` you can also run bloopInstall with an additional parameter:

> `mvn ch.epfl.scala:maven-bloop_2.10:1.3.0-RC1:bloopInstall -DdownloadSources=true -DaddScalacArgs=-Xplugin:@@semantic_DB_plugin_jar@@|-P:semanticdb:synthetics:on|-P:semanticdb:failures:warning|-P:semanticdb:sourceroot:/home/tgodzik/Documents/maven-test-repo|-Yrangepos|-Xplugin-require:semanticdb'`

Reviewer: ````suggestion
```
mvn ch.epfl.scala:maven-bloop_2.10:@@BLOOP_VERSION@@:bloopInstall -DdownloadSources=true -DaddScalacArgs=-Xplugin:/path/to/semanticdb-scalac.jar|-P:semanticdb:synthetics:on|-P:semanticdb:failures:warning|-P:semanticdb:sourceroot:/path/to/workspace|-Yrangepos|-Xplugin-require:semanticdb'
```
````

`-DaddScalacArgs` takes a string with additional scalac options separated by `|`. You would need to download the correct plugin first and manually replace `@@semantic_DB_plugin_jar@@` with the path to the semanticDB plugin jar.

> If you choose this option though you should disable automatic import, because it will override the changes.

Reviewer: ```suggestion
If you choose this option though you should select "Don't show again" when Metals prompts to import the build.
```

---
id: maven
title: maven
---

Maven is one of the most common build tools in the JVM ecosystem and it also allows for using scala through the [scala-maven-plugin](https://davidb.github.io/scala-maven-plugin/usage.html)

## Automatic installation

The first time you open Metals in a new workspace it prompts you to import the
build. Select "Import build" to start automatic installation.

This will create all the needed bloop files, however there will be a warning, since the semanticDB plugin was not added. Most features should work without it, however some require semanticDB files to be provided alongside compiled data. To do that we need a couple of steps that are explained in the manual instalation section.

## Manual instalation

First, add the following in the project's `pom.xml` `<dependencies>` section:

```xml
    <dependency>
      <groupId>org.scalameta</groupId>
      <artifactId>semanticdb-scalac_${scala.version}</artifactId>

>       <version>4.1.4</version>

Reviewer: ```suggestion
      <version>@@SCALAMETA_VERSION@@</version>
```

    </dependency>
```
`4.1.4` version can be replaced with whatever is the newest version.

Now, run `mvn compile`, which will download the needed scalac plugin. The jar should be available in your repository, which usually is `<user_dir>/.m2`.

Next, we need to add a couple of options to the compiler, that's why we will have to modify the configuration section of the plugin:

```xml
  <plugin>
    <groupId>net.alchim31.maven</groupId>
    <artifactId>scala-maven-plugin</artifactId>
    <version>4.0.2</version>
    <configuration>
      <args>

>         <arg>-Xplugin:@@replace_with_semanticDB_path@@</arg>

Reviewer: Using the absolute local path seems fragile, is there a Scala maven plugin that makes it easier to add a compiler plugin?

Reviewer: scalor-maven-plugin: https://random-maven.github.io/scalor-maven-plugin/2.12/eclipse-config-mojo.html#definePluginList

scala-maven-plugin: http://davidb.github.io/scala-maven-plugin/example_scalac_plugins.html (see "using a scalac plugin"

Author: Och alright, I haven't noticed this solution. That makes it much easier now.

        <arg>-P:semanticdb:synthetics:on</arg>
        <arg>-P:semanticdb:failures:warning</arg>

>         <arg>-P:semanticdb:sourceroot:${project.basedir}</arg>

Reviewer: Is `project.basedir` the root directory of the build or the base directory of the project? I'm wondering if there is actually such as thing as `baseDirectory.in(ThisBuild).value` in Maven 🤔 We might need built-in support for automatically setting the sourceroot in Bloop, cc/ @jvican

Author: `project.basedir: The directory that the current project resides in.`

Reviewer: For most common Maven builds I suspect we need the sourceroot to match the parent of `project.basedir`.

Author: Plugin is usually defined in the root project, so this should point to the correct location.

Reviewer: Have you double checked? It might be worth adding a test case to confirm

Author: Changed to `maven.multiModuleProjectDirectory` - that seems to be the root dir for all of the projects. Checked it and looks to be working.
